ksqldb-engine/src/test/java/io/confluent/ksql/query/PullQueryWriteStreamTest.java

```diff
@@ -0,0 +1,166 @@
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+
+import com.google.common.collect.ImmutableList;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.pull.StreamedRowTranslator;
+import io.confluent.ksql.rest.entity.StreamedRow;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.util.KeyValueMetadata;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PullQueryWriteStreamTest {
+
+  private static final QueryId TEST_ID = new QueryId("test");
+  private static final int TEST_TIMEOUT_SECONDS = 60;
+  private static final int LIMIT = 5;
+
+  @Rule
+  public final Timeout timeout = Timeout.seconds(TEST_TIMEOUT_SECONDS);
+
+  @Mock private LogicalSchema schema;
+  private PullQueryWriteStream writeStream;
+  private ScheduledExecutorService executorService;
+
+  @Before
+  public void setUp() {
+    final StreamedRowTranslator translator = new StreamedRowTranslator(schema, Optional.empty());
+    writeStream = new PullQueryWriteStream(OptionalInt.of(LIMIT), translator);
+    writeStream.write(ImmutableList.of(StreamedRow.header(TEST_ID, schema)));
+
+    executorService = Executors.newSingleThreadScheduledExecutor();
+  }
+
+  @After
+  public void tearDown() {
+    executorService.shutdown();
+  }
+
+  @Test
+  public void shouldAcceptWrites() {
+    // When:
+    writeStream.write(getData(1));
+
+    // Then:
+    assertThat(writeStream.getTotalRowsQueued(), is(1));
+    assertThat(writeStream.size(), is(1));
+  }
+
+  @Test
+  public void shouldPoll() {
+    // Given:
+    final AtomicBoolean written = new AtomicBoolean(false);
+    writeStream.write(getData(1), ignored -> written.set(true));
+
+    // When:
+    final KeyValueMetadata<List<?>, GenericRow> row = writeStream.poll();
+
+    // Then:
+    assertThat(row.getKeyValue().value().get(0), is(0));
+    assertThat("expected written callback to be called", written.get());
+    assertThat("expected writeStream to be empty after poll", writeStream.isEmpty());
+  }
+
+  @Test
+  public void shouldHandleLimit() {
+    // Given:
+    final AtomicBoolean limit = new AtomicBoolean(false);
+    final AtomicBoolean end = new AtomicBoolean(false);
+
+    writeStream.setLimitHandler(() -> limit.set(true));
+    writeStream.setCompletionHandler(() -> end.set(true));
+
+    // When:
+    writeStream.write(getData(LIMIT + 1));
```

Comment:
I'm still getting oriented... what happens to the extra data?  

Does it get dropped or just blocked?

If blocked, should there be a test to show that it gets unblocked and read?

Reply:
It just gets dropped if it hits the hard limit, which is what happens here. If it hits the soft limit it just goes through.